## Libraries and Loading

In [79]:
import pandas as pd
import chardet
import matplotlib.pyplot as plt
from collections import Counter
import google.generativeai as genai

In [35]:
Complete_data = r"C:\Frank\UoT 2024-2025\MIE8888 Project\Data\Complete Data\21_treatements_backup_joined_patients_procedure_na_removed.csv"
Patient_Procedure = r"C:\Frank\UoT 2024-2025\MIE8888 Project\Data\Complete Data\21_treatements_backup_joined_patients_procedure_grouped.csv"
rule_2 = r"C:\Frank\UoT 2024-2025\MIE8888 Project\Data\Complete Data\rules_2.csv"

## Encoding Type Helper Function

In [36]:
# with open(Complete_data, 'rb') as f:
#     raw_data = f.read()
#     result = chardet.detect(raw_data)
#     encoding = result['encoding']
# print(f"Detected encoding: {encoding}")

In [37]:
# with open(Patient_Procedure, 'rb') as f:
#     raw_data = f.read()
#     result = chardet.detect(raw_data)
#     encoding2 = result['encoding']
# print(f"Detected encoding: {encoding}")

## Temporarily remove empty columns

In [38]:
df = pd.read_csv(Complete_data, encoding="Windows-1252")
cleaned_df = df.dropna(axis=1, how='all')
cleaned_column_names = cleaned_df.columns.tolist()
print(cleaned_column_names)
print(cleaned_df.shape[1])

original_column_names = df.columns.tolist()
dropped_columns = [col for col in original_column_names if col not in cleaned_column_names]
print(dropped_columns)
print(len(dropped_columns))

C:\Users\frank\AppData\Local\Temp\ipykernel_12124\2635263723.py:1: DtypeWarning: Columns (1,4,10,16,52,85,86,91) have mixed types. Specify dtype option on import or set low_memory=False.
  df = pd.read_csv(Complete_data, encoding="Windows-1252")


['id_x_x', 'sikka_sub_task_x_x', 'sikka_sub_task_operation_x_x', 'name_x', 'office_x_x', 'patient', 'guarantor_x', 'provider_x', 'description', 'state_x', 'visible_to_patient', 'amount', 'estimated_cost', 'insurance_coverage', 'tooth_from', 'primary_insurance_estimate', 'secondary_insurance_estimate', 'plan_sr_no', 'entry_date', 'insurance_payment', 'treatment_plan_status', 'procedure_date', 'priority', 'approved', 'from_pms_x_x', 'id_y_x', 'from_pms_y_x', 'sikka_sub_task_y_x', 'sikka_sub_task_operation_y_x', 'uuid', 'patient_id', 'guarantor_y', 'office_y_x', 'fullname', 'first_name', 'middle_name', 'last_name', 'birth_date', 'address', 'city', 'state_y', 'first_visit', 'last_visit', 'email', 'phone', 'zip_code', 'practice', 'sikka_provider_id', 'primary_insurance_company_id', 'profile_picture', 'date_joined', 'zip_code_flag', 'address_comp', 'fetched_zip_code', 'fetched_zip_code_flag', 'zip_code_verified', 'final_zip_code', 'LATITUDE', 'LONGITUDE', 'id_x_y', 'from_pms_x_y', 'sikka_sub

## Task 1: Focus on the treatment processing per day, consider the ones with more than 7 treatment per day.

In [39]:
df2 = pd.read_csv(Patient_Procedure)
filtered_df = df2[df2['procedure_code_y_len'] >= 7]
unnormal_case = filtered_df[['procedure_code_y', 'procedure_code_y_len']]
unnormal_case.head()

procedure_code_y  procedure_code_y_len
939     [1203, 2142, 12101, 43412, 62501, 67211, 99111]                     7
1458  [1103, 1103, 2113, 2113, 2144, 2144, 2601, 260...                    10
1475     [1103, 2142, 2601, 11112, 11117, 12101, 99600]                     7
2810      [1103, 2112, 2142, 2601, 12101, 43412, 43413]                     7
3156  [1202, 2114, 11101, 11113, 12101, 25531, 27602...                     9

In [40]:
case_counts = Counter(unnormal_case['procedure_code_y'])
case_summary = pd.DataFrame.from_dict(case_counts, orient='index', columns=['Count'])
case_summary.head()

Count
[1203, 2142, 12101, 43412, 62501, 67211, 99111]         1
[1103, 1103, 2113, 2113, 2144, 2144, 2601, 2601...      1
[1103, 2142, 2601, 11112, 11117, 12101, 99600]          1
[1103, 2112, 2142, 2601, 12101, 43412, 43413]           1
[1202, 2114, 11101, 11113, 12101, 25531, 27602,...      1

In [41]:
# Consider the count of less than 10 as ramdom occurrence
case_summary = case_summary[case_summary['Count'] >= 10]
case_summary.head()

Count
[1202, 2144, 2144, 11107, 11107, 11113, 11113]         23
[1103, 2112, 2144, 2601, 11103, 12101, 99700]          41
[1202, 1202, 11107, 11107, 11113, 11113, 12111,...     11
[1202, 1202, 2144, 2144, 11107, 11107, 11113, 1...     29
[1202, 1202, 2144, 2144, 11107, 11107, 11113, 1...     11

In [42]:
# Get the procedure code for special case and use them to find the treatment description in full dataset
unique_numbers = set()
for idx in case_summary.index:
    numbers = eval(idx) if isinstance(idx, str) else idx
    unique_numbers.update(numbers)
unique_numbers = sorted(unique_numbers)
print("Unique Codes", unique_numbers)


Unique Codes [1103, 1202, 2112, 2142, 2144, 2601, 11101, 11103, 11107, 11111, 11113, 11117, 12101, 12111, 96103, 99700]


In [43]:
special_descriptions = []
for i in unique_numbers:
    matching_rows = df[df['procedure_code_y'] == i]
    if not matching_rows.empty:
        special_descriptions.append({
            'procedure_code_y': i,
            'special_code_description': matching_rows['procedure_code_description'].iloc[0],  # Take the first match
            'count': len(matching_rows)
        })
summary = pd.DataFrame(special_descriptions).sort_values('count', ascending=False)
summary


procedure_code_y                           special_code_description  count
12             12101            !!!Fluoride Treatment Foam Gel or Rinse  25357
10             11113                                Scaling Three Units  22911
1               1202  Exam & Diagnosis Limited Oral Previous Patient...  21516
6              11101                         Polishing One Unit of Time  16154
11             11117                                   Scaling 1/2 Unit  15616
8              11107                         Polishing 1/2 Unit of Time  15010
3               2142            Radiographs Intraoral Bitewing-2 Images  14629
4               2144          Radiographs Intraoral Bitewing - 4 Images  12256
5               2601                    Radiographs Panoramic - 1 Image  11676
9              11111                           Scaling One Unit of Time   7780
0               1103  Exam and Diagnosis Complete - Permanent Dentition   7569
2               2112         Radiographs Intraoral Periapical - 2 Image   6104
13             12111                                     Flouride Rinse   2895
15             99700                          !!!New Patient Exam-Adult   1288
7              11103                 !!!Prophylaxis Permanent Dentition   1043
14             96103           Dispensing Non Emergency Prescription +E    897

In [44]:
code_to_description = dict(zip(summary['procedure_code_y'], summary['special_code_description']))
case_summary['Description'] = case_summary.index.map(lambda codes: ", ".join([code_to_description[int(code)] for code in eval(codes) if int(code) in code_to_description]))
case_summary.sort_values('Count', ascending=False)

Count  \
[1103, 2112, 2144, 2601, 11103, 12101, 99700]          41   
[1202, 1202, 2144, 2144, 11107, 11107, 11113, 1...     29   
[1202, 2144, 2144, 11107, 11107, 11113, 11113]         23   
[1103, 2112, 2142, 2601, 11103, 12101, 99700]          20   
[1103, 2112, 2144, 2601, 11111, 96103, 99700]          16   
[1202, 2142, 2601, 11101, 11111, 11117, 12101]         14   
[1202, 1202, 11107, 11107, 11113, 11113, 12111,...     11   
[1202, 1202, 2144, 2144, 11107, 11107, 11113, 1...     11   

                                                                                          Description  
[1103, 2112, 2144, 2601, 11103, 12101, 99700]       Exam and Diagnosis Complete - Permanent Dentit...  
[1202, 1202, 2144, 2144, 11107, 11107, 11113, 1...  Exam & Diagnosis Limited Oral Previous Patient...  
[1202, 2144, 2144, 11107, 11107, 11113, 11113]      Exam & Diagnosis Limited Oral Previous Patient...  
[1103, 2112, 2142, 2601, 11103, 12101, 99700]       Exam and Diagnosis Complete - Permanent Dentit...  
[1103, 2112, 2144, 2601, 11111, 96103, 99700]       Exam and Diagnosis Complete - Permanent Dentit...  
[1202, 2142, 2601, 11101, 11111, 11117, 12101]      Exam & Diagnosis Limited Oral Previous Patient...  
[1202, 1202, 11107, 11107, 11113, 11113, 12111,...  Exam & Diagnosis Limited Oral Previous Patient...  
[1202, 1202, 2144, 2144, 11107, 11107, 11113, 1...  Exam & Diagnosis Limited Oral Previous Patient...

- Above is one table that shows the **unnormal treatment groups** (more than 7 treatments applied to one patient one day) that have happened more than 10 times in our data.  
- This table includes the description of what the treatment group as well as the exact times these treatment group appeared.
- These groups may reflect specific orders of the treatment group that doctors already knew;  
- But some are possibly not, and we can use these groups for the use of next-step-treatment prediction reference

---

## Task 2 Use aprior algorithm on co-occurrence detection

In [45]:
# Load the data containing 
df_r = pd.read_csv(rule_2)

## Understand rule_2 csv
Support: 0.002914619: The rule applies to 0.29% of transactions.  
Confidence: 0.559360731: If item 62501 is purchased, there’s a 55.94% chance that 67211 and 99111 will also be purchased.  
Lift: 182.599: The consequent is 182.6 times more likely to occur given the antecedent, compared to random chance.  

- The 231 rows in data rule_2 has lift > 3 means the consequent is at least 3 times more likely to occur given the antecedent, compared to random chance.
- Confidence all greater than 50% means that if antecedent is purchased, higehr than 50% of the chance that the consequent will also be purchased.
- A high support value for an itemset indicates that the itemset occurs in a large proportion of the transactions.


## Define what High & Low

High support indicates that an itemset is frequently purchased, while high lift shows a strong association between items that occur together much more often than by chance  
Low support indicates that an itemset is less frequently purchased, while low lift shows a weak association between items that occur together much more often than by chance

- H_Support + H_Lift: Very popular treatment group with consequent frequently happened
- H_Support + L_Lift: Very popular treatment group with consequent less frequently happened
- L_Support + H_Lift: common treatment group with consequent frequently happened
- L_Support + L_Lift: common treatment group with consequent less frequently happened


In [65]:
support_threshold = 0.005
lift_threshold = 30.0

def classify_group(row):
    if row['support'] >= support_threshold and row['lift'] >= lift_threshold:
        return 'H_Support_H_Lift'  # Very popular with frequent consequent
    elif row['support'] >= support_threshold and row['lift'] < lift_threshold:
        return 'H_Support_L_Lift'  # Very popular with less frequent consequent
    elif row['support'] < support_threshold and row['lift'] >= lift_threshold:
        return 'L_Support_H_Lift'  # Common with frequent consequent
    else:
        return 'L_Support_L_Lift'  # Common with less frequent consequent

df_r['Group'] = df_r.apply(classify_group, axis=1)
df_r['Treatment_Group'] = df_r['antecedents'].str.findall(r'\d+').apply(lambda x: [int(i) for i in x])
df_r['Next_Treatment_Group'] = df_r['consequents'].str.findall(r'\d+').apply(lambda x: [int(i) for i in x])

code_to_description = dict(zip(df['procedure_code_y'], df['procedure_code_description']))
def codes_with_descriptions(codes):
    return [code_to_description.get(code, f'Unknown ({code})') for code in codes]
df_r['Treatment_Group_desc'] = df_r['Treatment_Group'].apply(codes_with_descriptions)
df_r['Next_Treatment_Group_desc'] = df_r['Next_Treatment_Group'].apply(codes_with_descriptions)
df_r


antecedents                consequents  antecedent support  \
0           frozenset({62501})  frozenset({67211, 99111})            0.005211   
1    frozenset({67211, 99111})         frozenset({62501})            0.003063   
2    frozenset({62501, 99111})         frozenset({67211})            0.003170   
3           frozenset({67211})  frozenset({62501, 99111})            0.005074   
4           frozenset({62501})         frozenset({67211})            0.005211   
..                         ...                        ...                 ...   
226    frozenset({1202, 2114})         frozenset({11113})            0.017333   
227  frozenset({11117, 11101})         frozenset({12101})            0.018660   
228   frozenset({1202, 11117})         frozenset({12101})            0.022865   
229   frozenset({11113, 1204})         frozenset({12101})            0.009987   
230   frozenset({11112, 1202})         frozenset({12101})            0.013651   

     consequent support   support  confidence        lift  representativity  \
0              0.003063  0.002915    0.559361  182.599237               1.0   
1              0.005211  0.002915    0.951456  182.599237               1.0   
2              0.005074  0.002915    0.919325  181.189929               1.0   
3              0.003170  0.002915    0.574443  181.189929               1.0   
4              0.005074  0.004735    0.908676  179.091158               1.0   
..                  ...       ...         ...         ...               ...   
226            0.131342  0.008744    0.504461    3.840814               1.0   
227            0.150757  0.010731    0.575072    3.814555               1.0   
228            0.150757  0.011771    0.514828    3.414950               1.0   
229            0.150757  0.005092    0.509827    3.381777               1.0   
230            0.150757  0.006840    0.501089    3.323817               1.0   

     leverage  conviction  zhangs_metric   jaccard  certainty  kulczynski  \
0    0.002899    2.262478       0.999733  0.543840   0.558007    0.755409   
1    0.002899   20.492661       0.997579  0.543840   0.951202    0.755409   
2    0.002899   12.332457       0.997644  0.546875   0.918913    0.746884   
3    0.002899    2.342412       0.999552  0.546875   0.573090    0.746884   
4    0.004708   10.894442       0.999625  0.853162   0.908210    0.920926   
..        ...         ...            ...       ...        ...         ...   
226  0.006467    1.752956       0.752685  0.062487   0.429535    0.285517   
227  0.007918    1.998556       0.751876  0.067621   0.499639    0.323125   
228  0.008324    1.750396       0.723718  0.072731   0.428701    0.296455   
229  0.003586    1.732538       0.711402  0.032712   0.422812    0.271801   
230  0.004782    1.702194       0.708817  0.043413   0.412523    0.273232   

                Group Treatment_Group Next_Treatment_Group  \
0    L_Support_H_Lift         [62501]       [67211, 99111]   
1    L_Support_H_Lift  [67211, 99111]              [62501]   
2    L_Support_H_Lift  [62501, 99111]              [67211]   
3    L_Support_H_Lift         [67211]       [62501, 99111]   
4    L_Support_H_Lift         [62501]              [67211]   
..                ...             ...                  ...   
226  H_Support_L_Lift    [1202, 2114]              [11113]   
227  H_Support_L_Lift  [11117, 11101]              [12101]   
228  H_Support_L_Lift   [1202, 11117]              [12101]   
229  H_Support_L_Lift   [11113, 1204]              [12101]   
230  H_Support_L_Lift   [11112, 1202]              [12101]   

                                  Treatment_Group_desc  \
0    [PonticPorcelain/Ceramic/Polymer Glass Fused t...   
1    [RetainerPorc/Ceramic/Polymer Fused to Metal B...   
2    [PonticPorcelain/Ceramic/Polymer Glass Fused t...   
3    [RetainerPorc/Ceramic/Polymer Fused to Metal B...   
4    [PonticPorcelain/Ceramic/Polymer Glass Fused t...   
..                                                 ...   
226  [Exam & Diagnosi

## Use API to consider if the co-occurence of two treatment groups are logical

In [94]:
# genai.configure(api_key="AIzaSyCHj1FJXEc4rhjSQ2hBGc7QyZtKIh6J-1s")
# def check_medical_logic(treatment, next_treatment):
#     prompt = f"Treatment: {treatment}\nNext Treatment: {next_treatment}\nDoes this sequence follow a logical medical treatment path? Reply with 'Yes' or 'No'."
#     try:
#         model = genai.GenerativeModel('gemini-1.5-flash')
#         response = model.generate_content(prompt)
#         answer = response.text.strip()

#         return "Yes" if "Yes" in answer else "No"
#     except Exception as e:
#         print(f"Error: {e}")
#         return "Error"

# df_r = df_r['Comment'] = df_r.apply(lambda row: check_medical_logic(row['Treatment_Group_desc'], row['Next_Treatment_Group_desc']),axis=1)

In [95]:
# H_Support + H_Lift
H_support_H_lift = df_r[df_r['Group'] == 'H_Support_H_Lift']
H_support_H_lift = H_support_H_lift[['Group', 'Treatment_Group', 'Next_Treatment_Group', 'support', 'lift', 'Treatment_Group_desc', 'Next_Treatment_Group_desc']]
H_support_H_lift.head()

Group Treatment_Group Next_Treatment_Group   support  \
7   H_Support_H_Lift         [49510]              [49421]  0.007804   
8   H_Support_H_Lift         [49421]              [49510]  0.007804   
14  H_Support_H_Lift         [99700]   [2112, 2144, 1103]  0.005252   
18  H_Support_H_Lift         [99700]   [2144, 2601, 1103]  0.005383   
19  H_Support_H_Lift   [2112, 99700]         [2144, 1103]  0.005252   

         lift                               Treatment_Group_desc  \
7   71.043249                                  [Continuing Care]   
8   71.043249                                          [Hygiene]   
14  50.351599                        [!!!New Patient Exam-Adult]   
18  44.275259                        [!!!New Patient Exam-Adult]   
19  44.026720  [Radiographs Intraoral Periapical - 2 Image, !...   

                            Next_Treatment_Group_desc  
7                                           [Hygiene]  
8                                   [Continuing Care]  
14  [Radiographs Intraoral Periapical - 2 Image, R...  
18  [Radiographs Intraoral Bitewing - 4 Images, Ra...  
19  [Radiographs Intraoral Bitewing - 4 Images, Ex...

In [59]:
# H_Support + L_Lift
H_support_L_lift = df_r[df_r['Group'] == 'H_Support_L_Lift']
H_support_L_lift = H_support_L_lift[['Group', 'Treatment_Group', 'Next_Treatment_Group', 'support', 'lift', 'Treatment_Group_desc', 'Next_Treatment_Group_desc']]
H_support_L_lift.head()

Group Treatment_Group Next_Treatment_Group   support  \
40  H_Support_L_Lift   [2144, 99700]         [2601, 1103]  0.005383   
41  H_Support_L_Lift   [2112, 99700]         [2601, 1103]  0.005371   
42  H_Support_L_Lift    [2112, 2601]         [2144, 1103]  0.011623   
43  H_Support_L_Lift    [2144, 1103]         [2112, 2601]  0.011623   
45  H_Support_L_Lift         [99700]         [2601, 1103]  0.006484   

         lift                               Treatment_Group_desc  \
40  26.982936  [Radiographs Intraoral Bitewing - 4 Images, !!...   
41  26.950094  [Radiographs Intraoral Periapical - 2 Image, !...   
42  26.491972  [Radiographs Intraoral Periapical - 2 Image, R...   
43  26.491972  [Radiographs Intraoral Bitewing - 4 Images, Ex...   
45  25.383367                        [!!!New Patient Exam-Adult]   

                            Next_Treatment_Group_desc  
40  [Radiographs Panoramic - 1 Image, Exam and Dia...  
41  [Radiographs Panoramic - 1 Image, Exam and Dia...  
42  [Radiographs Intraoral Bitewing - 4 Images, Ex...  
43  [Radiographs Intraoral Periapical - 2 Image, R...  
45  [Radiographs Panoramic - 1 Image, Exam and Dia...

In [58]:
# L_Support + H_Lift
L_support_H_lift = df_r[df_r['Group'] == 'L_Support_H_Lift']
L_support_H_lift = L_support_H_lift[['Group', 'Treatment_Group', 'Next_Treatment_Group', 'support', 'lift', 'Treatment_Group_desc', 'Next_Treatment_Group_desc']]
L_support_H_lift.head()

Group Treatment_Group Next_Treatment_Group   support  \
0  L_Support_H_Lift         [62501]       [67211, 99111]  0.002915   
1  L_Support_H_Lift  [67211, 99111]              [62501]  0.002915   
2  L_Support_H_Lift  [62501, 99111]              [67211]  0.002915   
3  L_Support_H_Lift         [67211]       [62501, 99111]  0.002915   
4  L_Support_H_Lift         [62501]              [67211]  0.004735   

         lift                               Treatment_Group_desc  \
0  182.599237  [PonticPorcelain/Ceramic/Polymer Glass Fused t...   
1  182.599237  [RetainerPorc/Ceramic/Polymer Fused to Metal B...   
2  181.189929  [PonticPorcelain/Ceramic/Polymer Glass Fused t...   
3  181.189929  [RetainerPorc/Ceramic/Polymer Fused to Metal B...   
4  179.091158  [PonticPorcelain/Ceramic/Polymer Glass Fused t...   

                           Next_Treatment_Group_desc  
0  [RetainerPorc/Ceramic/Polymer Fused to Metal B...  
1  [PonticPorcelain/Ceramic/Polymer Glass Fused t...  
2  [RetainerPorc/Ceramic/Polymer Fused to Metal B...  
3  [PonticPorcelain/Ceramic/Polymer Glass Fused t...  
4  [RetainerPorc/Ceramic/Polymer Fused to Metal B...

In [57]:
# L_Support + L_Lift
L_support_L_lift = df_r[df_r['Group'] == 'L_Support_L_Lift']
L_support_L_lift = L_support_L_lift[['Group', 'Treatment_Group', 'Next_Treatment_Group', 'support', 'lift', 'Treatment_Group_desc', 'Next_Treatment_Group_desc']]
L_support_L_lift.head()

Group            Treatment_Group Next_Treatment_Group  \
39  L_Support_L_Lift        [2112, 99700, 2144]         [2601, 1103]   
44  L_Support_L_Lift                    [14612]              [99111]   
48  L_Support_L_Lift  [2144, 2601, 99700, 1103]               [2112]   
49  L_Support_L_Lift        [2144, 2601, 99700]               [2112]   
54  L_Support_L_Lift  [2112, 2601, 99700, 2144]               [1103]   

     support       lift                               Treatment_Group_desc  \
39  0.004824  27.301201  [Radiographs Intraoral Periapical - 2 Image, !...   
44  0.003533  25.651950             [Appliances Periodontal Mandibular +L]   
48  0.004824  24.897706  [Radiographs Intraoral Bitewing - 4 Images, Ra...   
49  0.004866  24.865327  [Radiographs Intraoral Bitewing - 4 Images, Ra...   
54  0.004824  22.235771  [Radiographs Intraoral Periapical - 2 Image, R...   

                            Next_Treatment_Group_desc  
39  [Radiographs Panoramic - 1 Image, Exam and Dia...  
44            [Commercial Laboratory Procedures (+L)]  
48       [Radiographs Intraoral Periapical - 2 Image]  
49       [Radiographs Intraoral Periapical - 2 Image]  
54  [Exam and Diagnosis Complete - Permanent Denti...